File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-0.py
class Solution:
    def largestRectangleArea(self, height: List[int]) -> int:
        n = len(height)
        def nearestLowLeft():
            left = []
            value = []
            index = []
            for i in range(n):
                if len(value) != 0 and value[-1] >= height[i]:
                    value.pop()
                    index.pop()
                    
                if len(value) == 0: 
                    left.append(-1) 
                else: 
                    left.append(index[-1])
                    
                value.append(height[i])
                index.append(i)
            return left 
        
        def nearestLowRight():
            right = []
            value = []
            index = []
            for i in range(n-1, -1, -1):
                if len(value) != 0 and value[-1] >= height[i]:
                    value.pop()
                    index.pop()
                    
                if len(value) == 0:
                    right.append(n)
                else: 
                    right.append(index[-1])
                    
                value.append(height[i])
                index.append(i)
            right.reverse()
            return right
        
        left = nearestLowLeft()
        right = nearestLowRight()
        
        maxArea = 0
        for i in range(n):
            area = (right[i] - left[i] - 1) * height[i]
            maxArea = max(maxArea, area)
        return maxArea

File: Data Structures & Algorithms/max-area-of-island/submission-0.cpp
class Solution {
public:
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int ROWS = grid.size();
        int COLS = grid[0].size();
        int max_area = 0;
        for (int i = 0; i < ROWS; i++)
        {
            for (int j = 0; j < COLS; j++)
            {
                if (grid[i][j] == 1)
                {
                    int c = 0;
                    dfs(i, j, ROWS, COLS, grid, c);
                    max_area = max(max_area, c);
                }
            }
        }
        return max_area;
    }

    void dfs(int r, int c, int rows, int cols, vector<vector<int>>& grid, int &area)
    {
        if (r < 0 || c < 0 || r >= rows || c >= cols || grid[r][c] != 1)
            return; 
        
        area += 1;
        grid[r][c] = 2;

        dfs(r+1, c, rows, cols, grid, area);
        dfs(r-1, c, rows, cols, grid, area);
        dfs(r, c+1, rows, cols, grid, area);
        dfs(r, c-1, rows, cols, grid, area);
    }

};
